### ISSUE TITLE
Не рендерятся сниппеты при выключеном «Enable filters»

### ISSUE THREAD

**Issue by neilsagitov** [evolution-cms/evolution#977]
После обновления с 1.4.7, если выключен «Enable filters», перестал парситься некэшируемый вызов сниппетов в чанках, просто преобразуется в кэшируемый [[...]]. Также, остаётся необработанным вызов сниппета, у которого в параметрах есть вложенный вызов сниппета.

**Comment by leonie12** [comment]
Same issue here.

**Comment by bossloper** [comment]
Yes I get this too, I assume it was by design(?) as this was the change by @dmi3yy : https://github.com/evolution-cms/evolution/commit/f04a856fcf3945992a24fee4e01957aadf909fbb

**Comment by dmi3yy** [comment]
Send plz sample of not rendered snippet i check

**Comment by bossloper** [comment]
This failed for me on 1.4.8 (ok on 1.4.7) unless 'enable filters' is set to yes:
```
<header class="headerimage hidden-xs"  [[if? &is=`[*headerImage*]:!empty` &then=`style="background-image:url([[phpthumb? &input=`[*headerImage*]` &options=`w=1920,h=300,zc=far,q=60`]])"`]] >
```

**Comment by dmi3yy** [comment]
Check on 1.4.8 
<img width="547" alt="1" src="https://user-images.githubusercontent.com/669491/51472932-a05ecf00-1d83-11e9-8abf-1b7034c13415.png">
<img width="254" alt="2" src="https://user-images.githubusercontent.com/669491/51472940-a2c12900-1d83-11e9-9959-771ebab49a17.png">
<img width="285" alt="3" src="https://user-images.githubusercontent.com/669491/51472943-a48aec80-1d83-11e9-8c71-72b5470f2513.png">


have phx plugin ?

**Comment by dmi3yy** [comment]
test all combination 
[[ [[ 
[[ [! 
[! [[
[! [! 

all work correct with filters and without 

and check inside chunk all correct to

**Comment by bossloper** [comment]
no phx plugin.

My code actually has a further nested IF snippet that I think is causing the issue.

I created a very rough template in my default 1.4.8 install:

```<html>
	<body>
<p>test start</p>
<hr/>
[[if? &is=`[*hideTopImageBanner*]:empty` &then=`
<header class="headerimage hidden-xs"  [[if? &is=`[*headerImage*]:!empty` &then=`style="background-image:url([[phpthumb? &input=`[*headerImage*]` &options=`w=1920,h=300,zc=far,q=60`]])"`]] >
  <div class="container">
    &nbsp;
  </div>
</header>
`]]
<hr/>
<p>test end</p>
	</body>
</html>
```

At runtime the page source (without enable filters) is this:
```
<html>
	<body>
<p>test start</p>
<hr/>

<header class="headerimage hidden-xs"  [[if? &is=`:!empty` &then=`style="background-image:url([[phpthumb? &input=
<hr/>
<p>test end</p>
	</body>
</html>
```


If I make the outer snippet uncached then it is ok:

```
[!if? &is=`[*hideTopImageBanner*]:empty` &then=`
```

**Comment by dmi3yy** [comment]
`
Enable filters: [(enable_filter)] <br>
<hr> Work correct  
1. [[if? &is=`[*image*]:!empty` &then=`
	[[phpthumb? &input=`[*image*]` &options=`w=1920,h=300,zc=far,q=60`]]
`]] 
<hr> 3 and more nested work only if enable_filter = 1
2. [[if? &is=`[*alias*]:=:test1` &then=`
	[[if? &is=`[*image*]:!empty` &then=`
		[[phpthumb? &input=`[*image*]` &options=`w=1920,h=300,zc=far,q=60`]]
	`]] 
`]]

<hr> if put code with 2 level in IF correct @TPL will work 2 level:
3. [!if? &is=`[*alias*]:=:test1` &then=`@TPL:chunk`!]

<hr> if put code with 2 level nested incorect will not work:
4. [[if? &is=`[*alias*]:=:test1` &then=`{{chunk}}`]]

<hr>
Chunk: 
[[if? &is=`[*image*]:!empty` &then=`
	[[phpthumb? &input=`[*image*]` &options=`w=1920,h=300,zc=far,q=60`]]
`]] 

all this code correct work if enable_filter = 1. 
only 1 and 3 work if enable_filter = 0.  
`

**Comment by dmi3yy** [comment]
<img width="352" alt="12345" src="https://user-images.githubusercontent.com/669491/51491149-23008200-1db6-11e9-8bfc-576ae07d59a1.png">

If we use 2+ nested levels snippets all of this snippets run 2 time each.

**Comment by bossloper** [comment]
Yes correct I have option 2. from your list and it does not work on 1.4.8. Should it?

**Comment by dmi3yy** [comment]
Turn on enable_filter in setting and will work

**Comment by dmi3yy** [comment]
there are 2 options:
- 1 to return back to let the embedded snippets run 2 times and 2 times longer
- 2 check whether everything is ok and either correct the code or if everything is happy to switch the setting Enable filters

starting from version 1.4.8, there will only be a correction of errors and problems in security, so there will not be more similar problems on version 1.4

If need use 3 nested level need write new snippet and use it inside. Its best practice. Use 3 level nested snippets is bad practice 

Yes, I missed this point and did not document in the release.

**Comment by dmi3yy** [comment]
I writу about this in Release notes 
http://take.ms/dWFg5

**Comment by dmi3yy** [comment]
[Release Evolution CMS 1.4.8 · evolution-cmsevolution 2019-01-21 20-44-11.png.download.zip](https://github.com/evolution-cms/evolution/files/2780147/Release.Evolution.CMS.1.4.8.evolution-cmsevolution.2019-01-21.20-44-11.png.download.zip)

**Comment by dmi3yy** [comment]
<img width="413" alt="235" src="https://user-images.githubusercontent.com/669491/51493265-7cb87a80-1dbd-11e9-8a55-3db5430f9a90.png">

**Comment by bossloper** [comment]
... though my three level example may not be uncommon?

**Comment by dmi3yy** [comment]
your call is normal all 3-nested will not work

**Comment by bossloper** [comment]
My call is 3 nested.

**Comment by bossloper** [comment]
I think this can be closed.

**Comment by dmi3yy** [comment]
Try this and you see different: 
[[snippet? &param=`[[snippet]]`]]

`<?php
$modx->logEvent(123, 1, 'snippet1', 'snippet1');
`
after run we wait 2 logs

but before fix we see 4 logs ( 
<img width="167" alt="4" src="https://user-images.githubusercontent.com/669491/51494556-aa072780-1dc1-11e9-9810-4b1ddfa6d555.png">

if run this: 
[[snippet? &param=`[[snippet? &param=`[[snippet]]`]]`]] 
we wait 3 logs, but see 11: 
<img width="645" alt="11" src="https://user-images.githubusercontent.com/669491/51494646-09fdce00-1dc2-11e9-94f3-efeba7e560f8.png">

I think it good reason not use 3-level nested and not use filters.

**Comment by bossloper** [comment]
Totally agree, though in my example I am not sure I can reduce it to two levels... and in the real world the site is responsive enough with the 3 level call.
I think it is possible to take efficiency too far vs functionality, and I am quite ODC (Obsessive-compulsive disorder) if that translates!

Again as per my first comment if it is expected result for 1.4.8 then I think this is discussed and @neilsagitov issue is resolved?
The solution is 'enable filters' = yes if you have complex and inefficient nested snippet calls.

**Comment by neilsagitov** [comment]
> Totally agree, though in my example I am not sure I can reduce it to two levels...

To reduce to two levels use `@TPL` with chunk. 
For example, place first `then` to chunk `header_block`:

```
<header class="headerimage hidden-xs"  [[if? &is=`[*headerImage*]:!empty` &then=`style="background-image:url([[phpthumb? &input=`[*headerImage*]` &options=`w=1920,h=300,zc=far,q=60`]])"`]] >
  <div class="container">
    &nbsp;
  </div>
</header>
```

and call snippet:

```
[[if? &is=`[*hideTopImageBanner*]:empty` &then=`@TPL:header_block`]]
```

**Comment by pmfx** [comment]
@neilsagitov and it also prevents running snippet (phpthumb in this case) in `&then` parameter when condition result is false.

Remember, this will execute both snippet_1 and snippet_2, no matter what the `&is` result:
```
[[if? 
&is=`[*foo*]:!empty` 
&then=`[[snippet_1]]`
&else=`[[snippet_2]]`
]]
```
And this will execute only the one we need:
```
[[if? 
&is=`[*foo*]:!empty` 
&then=`@TPL:snippet_1_in_chunk`
&else=`@TPL:snippet_2_in_chunk`
]]
```
Huge performance difference in some cases.

**Comment by 64j** [comment]
@neilsagitov 

snippet **switch**

```php
<?php
$switch = !empty($switch) ? $switch : '';
$default = !empty($default) ? $default : '';
$out = '';
if (!empty($params)) {
	$out = $default;
	foreach($params as $k => $v) {
		$value = mb_substr($k, 5);
		if (mb_substr($k, 0, 5) === 'case:' && $value === $switch) {
			$out = $v;
			continue;
		}
	}
	return $out != '' ? $this->parseText($modx->getTpl($out)) : '';
}
```

example
```
[[switch?
&switch=`[*id*]`
&case:123=`@CODE or chunkName`
&case:456=`@CODE or chunkName`
&case:789=`@CODE or chunkName`
&default=`@CODE or chunkName`
]]
```
or
```
[[switch?
&switch=`[*longtitle*]`
&case:=`@CODE:<h1>[*pagetitle*]</h1>`
&default=`@CODE:<h1>[*longtitle*]</h1>`
]]
```

**Comment by bossloper** [comment]
Suggesting maybe the 'Enable Filter' system option help text should include extra info? e.g.:
"For best site performance do not enable unless nesting of snippets 3+ levels is required"

**Comment by Ronef** [comment]
I think, 3 levels is normal practice and very confused about it.

`Enable filter` setting affects to snippet calls levels 🤦‍♂️ It doesn't make sense!

And yes, i want to use 3, 4, 5 or 100500 levels. Why not? Performance is my business. It may be very simple site with 3 pages and all snippet calls are cashed. Why can't i use 3 level calls in this case? I can and i will be.

It looks like a hastily patched bug and unwillingness to solve the problem for real 🤔
